## Content

- Problem Statement
- Geometric Intuition of SVM
  - Hard Margin SVM
  
- Soft Margin SVM
  - Algebric Intuition of SVM
- Intution of Hinge Loss
- SVM Imbalance
  
- Code Implementation of Linear SVM




























---

### Problem Statement

Imagine youself as a Data Scientist at Google. </br>
You've been asked to come up with model to classify emails as either **Spam** or **Ham**.





---

### SVMs
- Popular in 2000's ( late 90s )
- Kernel SVM
- Theoritically they are best.
  - (In practice better algorithms exists)
- Less frequently used nowadays
- Challenging Maths

---

### Geometric intution behind SVM

The main idea behind SVM is to
- find the line or plane that can
- best seperate the given classes.

\
Let us try to understand this through a simple example.

Suppose,

We have some datapoints that belong to two different classes.
- +ve class
- -ve class

\
#### How do you divide the +ve samples from the -ve ones?

Using a **line/hyperplane**.

```
Hyperplanes can be defined as decision boundaries that help classify the datapoints.
```

But there can be multiple lines/hyperplanes.

\
#### So which line/hyperplane should we choose?

Imagine, </br>
We have two different **hyperplanes**.
- $\pi_1$ and $\pi_2$

\
<img src='https://drive.google.com/uc?id=1lFAoAj0_EimHorfMbhySFQLZlErMFq1W' height='500' width='400'>

\
#### Which hyperplane is better for separating the datapoints?

- Option A: $\pi_1$
- Option B: $\pi_2$

Answer: Hyperplane $\pi_2$

---

#### But how can we say that?

Because,
- the minimum distance of hyperplane $\pi_2$
- from the nearest existing data points is large.

\
Intuitively,
- If we look at the hyperplane $\pi_1$
- and we draw two hyperplanes parallel to $\pi_1$ from the closest +ve & -ve datapoint to $\pi_1$
- we can see the gap between these two parallel drawn hyperplanes.

Let's call this gap as **margin_1** wrt hyperplane $\pi_1$.

\
Similarly,
- we can get **margin_2** wrt hyperplane $\pi_2$.

Now we can see that,
- margin_2 is much larger than margin_1.

Hence,
- we pick that hyperplane $\pi$ that results in largest margin.

---

#### Why do we need the margin to be large?

The hyperplane is drawn with a view of putting in
- the **widest street** that separates the +ve & -ve samples.

\
Therefore,
- the larger the margin,
- the better the separation.

\
Such classifiers where,
- we want the margin to be as large as possible
- are called **margin-maximizing classifiers**.

\
```
NOTE:
Distances from datapoints are measured perpendicular to any of the hyperplanes.
```

```
QUIZ 1:

Q. What do you mean by generalization in terms of SVM?

A. How far the hyperplane is from the training datapoints
B. How accurately the SVM can predict outcomes for unseen data
C. How accurately the SVM classifies training datapoints

ANS:
B. How accurately the SVM can predict outcomes for unseen data
```

---

#### How do we define these hyperplanes?

Recall from regression models, </br>
we define a hyperplane as:
- $\pi : w^Tx+b = 0$

\
Let's assume, </br>
the parallel hyperplanes are defined as:
- For $\pi^+$,
  - $w^Tx+b = 1$
- For  $\pi^-$,
  - $w^Tx+b = -1$

\
<img src='https://drive.google.com/uc?id=10DaNlGobo-V7GOu9Qr2PP2D_UBnnqe1Y' height='500' width='450'>



---

Let's summarize everything.

\
We have a bunch of +ve & -ve datapoints.

- Let's say $\pi$ is a margin-maximizing hyperplane,

\
  - $\pi^+$ is the positive hyperplane parallel to $\pi$ and
    - touching the closest postive points to the $\pi$
  - $\pi^-$ is the +ve hyerplane  parallel to $\pi$ and
    - touching the closest -ve points to the $\pi$

\
- $d$ is margin i.e., distance between $\pi^+$ and $\pi^-$
- We want this margin to be as large as possible.

---

#### What will be the length of the margin if $||w|| \neq 1$?

Recall from linear algebra,

\
If we rearrange the equations for $\pi^+$ and $\pi^-$,
- $w^Tx+b = 1 $ </br>
  $\Rightarrow$ $w^Tx+b -1 = 0 $
- $w^Tx+b = -1 $ </br>
  $\Rightarrow$ $w^Tx+b+1 = 0 $

\
If measured from the origin,
- the distance of hyperplane $\pi^+$ can be defined as
  - $ d(0, \pi^+) = \frac{b-1}{||w||}$

- the distance of hyperplane $\pi^-$ can be defined as
  - $ d(0, \pi^-) = \frac{b+1}{||w||}$

\
Hence,
- the distance between the two hyperplanes will be
- $\frac{b+1}{||w||} - \frac{b-1}{||w||}$

\
$\Rightarrow$ Margin i.e., </br> $d( \pi^+,\pi^-) = \frac{2}{||w||}$

\
<img src='https://drive.google.com/uc?id=1l1mOyJbXlB5vHnQISpy_EmZqJvTMuH7E' height='450' width='450'>

Now, our goal is
- to maximize this margin $\frac{2}{||w||}$ in order to
- obtain the best possible separation between the two classes.

\
#### What will be the parameters for the margin $\frac{2}{||w||}$?

- Weight ($w$)
- Constant ($b$)

---

### What if we define the hyperplanes like this?
- $\pi^+$ as $w^Tx+b=k$
- $\pi^-$ as $w^Tx+b=-k$

\
If we take $k=10$ then,
- $\pi^+:w^Tx+b=10$
- $\pi^-: w^Tx+b=-10$

\
Hence, </br>
we get our margin as:
- $\frac{20}{||w||}$

\
Now we will maximize this margin on
- weight ($w$) and
- constant ($b$)

\
<img src='https://drive.google.com/uc?id=1sdtB4cTIJ4-bEWa_8cGnz-AbCLAm9SbU' height='450' width='450'>

\
Since the margin does not change,
- with respect to $w$ and $b$

\
<img src='https://drive.google.com/uc?id=1SP2ZwDUV3AEzDVHAGzfraPYMcwrs9Ppf' height='450' width='450'>

It doesn't matter if we take
- $+1,-1$ or $k,-k$

Hence,
- for mathematical simplicity,
- we take $+1, -1$

```
QUIZ 2:

Q. If,
- π+ : w^T * x + b = 40
- π- : w^T * x + b = -50
then margin will be:

A. 10/||w||
B. 40/||w||
C. 50/||w||
D. 90/||w||

ANS:
D. 90/||w||
```

---

### <b> SVM Demo </b>

https://jgreitemann.github.io/svm-demo

<img src='https://drive.google.com/uc?id=149Xs-dDaEhXH8m90fiSlT0SPlUchUo2T' height='400' width='650'>

### Hard Margin SVM

Now, as we discussed
- our aim is to maximize the margin.

\
#### So how can we perform optimization here?

Let's look at an example to understand this.

\
What will be the value of a +ve datapoint which lies on the hyperplane $\pi^+$?
- 1 ,
- since $\pi^+: w^Tx+b=1$

```
Case 1:
```
What will be the value of the +ve datapoints which lie beyond the hyperplane $\pi^+$?
- greater than 1 ,
- hence $\pi^+: w^Tx+b > 1$

\
Similarly,

```
Case 2:
```
What will be the value of the -ve datapoints which lie beyond the hyperplane $\pi^-$ ?
- less than -1 ,
- hence $\pi^+: w^Tx+b < -1$




---

Now, </br>
for mathematical convenience </br>
we introduce a term $y_i$ such that:

- $y_i$ = +1 for +ve samples
- $y_i$ = -1 for -ve samples

\
Therefore,
- $\pi^+: y_i(w^Tx+b) > 1$
- $\pi^+: y_i(w^Tx+b) < -1$

\
We can club both the cases and say,
- we maximize the margin such that
- for all n samples,
- our $(w^Tx_i + b)y_i \geq 1 $

\
<img src='https://drive.google.com/uc?id=1TdMTmU8OeS-jZss9Io9VD3LKKTcfQG9W' height='500' width='400'>

\
#### But how does this work?

- For +ve samples,
  - $y_i$ = +1
  - $(w^Tx_i + b)$ will be +ve since it will be $\geq$ 1
  - hence, (+ve) multiplied by (+ve) makes positive.
- For -ve samples,
  - $y_i$ = -1
  - $(w^Tx_i + b)$ will be -ve since it will be $\leq$ 1
  - hence, (-ve) multiplied by (-ve) makes positive.

Example -

<img src='https://drive.google.com/uc?id=1VgUbWlosavzPc9ftYrdt0JntpwZP6c1_' height='400' width='750'>


---

#### But why the $(w^Tx_i + b)y_i \geq 1$ constraint?

Our goal is to
- seperate the two classes completely
- with margin as maximum as possible.

\
With this constraint:
- all +ve datapoints should lie beyond $\pi^+$
- all -ve datapoints should lie beyond $\pi^-$

\
Thus,
- it expects the hyperplane to have zero errors since,
- it does not want any datatpoint to belong
- to the wrong side of the parallel hyperplanes.

To summarise, </br>

If we strictly impose that:
- all instances must be off the street (margin)
- and on the correct side,
- then this is called **Hard Margin classification**.


---

#### When would Linear SVM with Hard Margin fail?

<img src='https://drive.google.com/uc?id=1Sieivv5mv2kFdKdXdjcc51bLTybnq0iC' height='400' width='500'>



```
QUIZ 3:

Q. What do you mean by a hard margin?

A. The SVM allows no error in classification.
B. The SVM allows some error in classification.
C. The SVM allows high error in classification.

ANS:
A) The SVM allows very low error in classification.
```
```
Explanation:

A hard margin means that an SVM is very rigid in classification and
tries to work extremely well on the training set causing overfitting.
```




---

### Soft Margin SVM

#### What if data is not perfectly linearly seperable?

Imagine a dataset where,
- some data points are on the wrong side.

This is what we call **almost linearly separable** data.

\
<img src='https://drive.google.com/uc?id=15l_dsx_MUv7PB8YbQPJPWW0RQj2lueI3' height='420' width='475'>

\
#### How to account for these data points?

Imagine,
- A $+ve$ labelled data point $ \ x_1 $
- at 0.5 unit distance in between $ \pi \ and \ \pi^+ $

\
#### What will be the value of $ y_i(w^Tx_i+b) $ for $x_1$?

- $\ y_1(w^Tx_1+b)$ = 0.5 = 1 - 0.5,
- where 0.5 is error $ \zeta_1 $

\
#### Does this equation $(w^Tx_i+b)y_i >= 1 - \zeta_i$ holds true for -ve points too?

**Yes**

Imagine,

- A $-ve$ labelled data point $ \ x_2 $
- at 0.5 unit distance in between $ \pi \ and \ \pi^- $

\
#### What will be the value of $ y_i(w^Tx_i+b) $ for $x_2$?

- $\ y_2(w^Tx_2+b)$ = 0.5 = 1 - 0.5,
- where 0.5 is $ \zeta_2 $

\
<img src='https://drive.google.com/uc?id=19Jy5DJE9zZMqmxos8FWhH-RJdonqfIi9' height='450' width='400'>

\
So,
- $ \zeta_i = 0$ for all correctly placed points.
- $ \zeta_i > 0$ for all incorrectly placed points.

\
Now, our optimization problem becomes:

- $ max \ \frac{2}{||w||} $ i.e., the margin
- along with minimizing error $ \zeta_i's $

because we're try to get the best possible classificaton.

\
#### Can we think of another way to write this?

Reciprocating above equation,

- $ min \ \frac{||w||}{2} $ with $ \zeta_i's $



---

#### What do you think our goal here is ?

- We want to maximize the margin
- and minimize data points having $ \zeta_i > 0$,
- so we should minimize the error $ \zeta_i's $.

\
Now the optimization function changes to:
- $ min_{w,b} \ \frac{||w||}{2} + C \sum_{i=1}^N \zeta_i$

- such that $(w^Tx_i+b)y_i >= 1 - \zeta_i$

- for all $i : 1 \rightarrow N$

-  $ \zeta_i >= 0 $.

\
This is called as **SVM with soft margin** which we use when we have **almost linearly seperable** data.

---

#### What's the use of $C$ here?

$C$ is a hyperparameter.

\
It controls whether we have to focus on
- maximizing the margin or
- minimizig the errors $ \zeta_i's $

\
#### What if the value of $C$ becomes zero?

As $C \downarrow$
- more imp will be given to maximize the margin
- so the model Underfits.

\
<img src='https://drive.google.com/uc?id=1-9SaALYoiCntg7xzQCqZQVLxuHUgsZ0L' height='400' width='500'>

\
#### What will happen if $C$ is very large?

As $C \uparrow$
- more imp will be given to the errors
- so the model Overfits.

\
<img src='https://drive.google.com/uc?id=1_KGPoD7bcYJcjMdPWTpRdhcflo6krjyc' height='400' width='500'>

\
Therefore,
- we need to find a balance here.



```
QUIZ 4:

Q. What would happen when you use very large value of C?

A. We can still classify training data correctly for given setting of hyperparameter C.
B. We cannot classify training data correctly for given setting of hyperparameter C.
C. Can’t say for sure

ANS:
A. We can still classify training data correctly for given setting of hyperparameter C.
```
```
Explanation:

For large values of C,
- the penalty for misclassifying points is very high,
- so the decision boundary will perfectly separate the data.


---

### Algebric intuition behind SVM

We saw that,
- Soft Margin SVMs are defined as:
- $min$ <sub> $(w,b)$ </sub> $\frac{||w||}{2}$ + $C\frac{1}{N}$$\sum_{i=1}^N \zeta_i$

\
If we look closely,
- the term $\frac{1}{N}$$\sum_{i=1}^N \zeta_i$
- is the error which we try to minimize.

\
Since $ζ_i$ will be non zero,
for data points which are
- either inside the margin
- or are misclassified.

\
**Note:** </br>
We refer to the term $\frac{1}{N}$$\sum_{i=1}^N \zeta_i$ as **Hinge Loss**.

---

If we now look,
- the term $\frac{||w||}{2}$

Geometrically,
- $\frac{||w||}{2}$ is used to have maximum margin.

Algebraically,
- $\frac{||w||}{2}$ is just $\frac{1}{2}$ of L2-Regularization.

\
Also, </br>
$C$ becomes analogous to our regularization hyperparameter (λ).

\
Therefore, </br>
we can interpret our Soft-Margin SVM as   
 - $C$ HingeLoss $+\frac{1}{2}$ L2Reg


---

### Intuition of Hinge Loss

\
<img src='https://drive.google.com/uc?id=1G0h0PNfMntRh7ncndEOZiVcNLBI13VKJ' height='400' width='550'>

\
Let's say
- we define our hyperplane $w^Tx+b$ as $f(x_i)$.
- and assume our x-axis to be:
 - $z_i = y_if(x_i)$

```
Case 1:
```
If $z_i \geq 1$
- or we can say $w^Tx+b \geq 1$

Then $\zeta_i = 0$
- i.e., the distance of $x_i$ from $\pi_+ = 0$
- meaning that the point lies beyond $\pi_+$

\
```
Case 2:
```
If $z_i = 0$
- or we can say $w^Tx+b = 0$

Then $\zeta_i = 1$
- i.e., the distance of $x_i$ from $\pi_+ = 1$
- meaning that the point lies on $\pi$

\
```
Case 3:
```
If $z_i = -1$
- or we can say $w^Tx+b = -1$

Then $\zeta_i = 2$
- i.e., the distance of $x_i$ from $\pi_+ = 2$
- meaning that the point lies on $\pi_-$

```
QUIZ 5:

Q. x1,x2,x3 are -ve datapoints which are 0.2, 3.0, 1.0 at unit distance from π-,
what will be the respective ζi?

A. 0.8, -2.0, 0.0
B. 0.2, 3.0, 1.0
C. 0.8, 2.0, 0.0

ANS:
B. 0.2, 3.0, 1.0
```

---

#### What can we conclude from the graph above?

1. Error cannot be negative.
  - $\zeta_i >= 0$
2. A/C to the constraint,
  - $y_i (w^Tx+b) > 1 - \zeta_i$

$~~~~~~~~~~~~ \Rightarrow z_i > 1 - \zeta_i$

$~~~~~~~~~~~~ \Rightarrow \zeta_i \geq 1- z_i$

\
**Note:**

As $z_i$ increases,
- $\zeta_i$ will reduce
- but it'll not go below zero.

As $z_i$ decreases,
- $\zeta_i$ will increases.

\
We can actually combine these two eqns into: </br>

**Hinge Loss** </br>
$ ~~~~~ \Rightarrow \zeta_i = max ~ (0, 1-z_i)$

---

### Comparison with Log Loss

<img src='https://drive.google.com/uc?id=1zpPvhQLQtp-I4FU4YwOkLZQJIG3cqk3H' height='400' width='550'>

\
#### What happens if $y_i ~ \epsilon ~ \{-1, +1\}$ for LogLoss?

- If, $y_i ~ \epsilon ~ \{-1, +1\}$

- Then, </br>
LogLoss
= $ ∑_{i=1}^{n} log( 1 + e^{(-z_i))} )$

$~~~~~~~~~~~~~~~~~~~~~$ = $∑_{i=1}^{n} log( 1 + e^{( -y_i (w^Tx_i + b))} )$

\
We will not be deriving how we get this equation.

---

### SVM Imbalance

#### Are SVMs affected by class imbalance?


If we look at optimization problem,

\
Only a few datapoints will be contributing to the hinge loss ($\zeta_i$).
- These points are called **Support Vectors**.


Hence, SVM will only be affected
- if there is imbalance in the no. of support vectors from each class.

\
#### Should we use SVM as the baseline model if we have imbalanced data?

Not necessarily.

Bcoz the balance in no. of support vectors from each class can't be guranteed.

\
**Note:** </br>
To handle imbalance in dataset,
- either use class weights
- or rebalance you data.

```
QUIZ 6:

Q. SVM will be impacted if there's an imbalance in the no. of datapoints belonging to each class?

A. True
B. False

ANS:
B. False
```
```
Explanation:

SVM will only be affected if there is imbalance in the number of support vectors from each class.
```

---

### Code implementation of Linear SVM

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from collections import Counter
from sklearn import feature_extraction, model_selection, naive_bayes, metrics, svm
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
import warnings

warnings.filterwarnings('ignore')
%matplotlib inline

In [ ]:
!gdown 1QViUZJ5UIBCgxB_qbOXTLs_2V48w7MWo

df = pd.read_csv('Spam_processed.csv', encoding='latin-1')
df.dropna(inplace = True)

Downloading...
From: https://drive.google.com/uc?id=1QViUZJ5UIBCgxB_qbOXTLs_2V48w7MWo
To: /content/Spam_processed.csv
100% 767k/767k [00:00<00:00, 5.50MB/s]


In [ ]:
df

type                                            message  \
0        0  Go until jurong point, crazy.. Available only ...   
1        0                      Ok lar... Joking wif u oni...   
2        1  Free entry in 2 a wkly comp to win FA Cup fina...   
3        0  U dun say so early hor... U c already then say...   
4        0  Nah I don't think he goes to usf, he lives aro...   
...    ...                                                ...   
5567     1  This is the 2nd time we have tried 2 contact u...   
5568     0              Will Ì_ b going to esplanade fr home?   
5569     0  Pity, * was in mood for that. So...any other s...   
5570     0  The guy did some bitching but I acted like i'd...   
5571     0                         Rofl. Its true to its name   

                                        cleaned_message  
0     go jurong point crazy available bugis n great ...  
1                               ok lar joking wif u oni  
2     free entry 2 wkly comp win fa cup final tkts 2...  
3                   u dun say early hor u c already say  
4             nah nt think goes usf lives around though  
...                                                 ...  
5567  2nd time tried 2 contact u u å750 pound prize ...  
5568                       ì_ b going esplanade fr home  
5569                              pity mood suggestions  
5570  guy bitching acted like interested buying some...  
5571                                     rofl true name  

[5565 rows x 3 columns]

- Performing train-test split
- with [CountVectorization](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)
- and StandardScaler.

In [ ]:
from sklearn.model_selection import train_test_split

df_X_train, df_X_test, y_train, y_test = train_test_split(df['cleaned_message'], df['type'],
                                                          test_size=0.25, random_state=47)
print([np.shape(df_X_train), np.shape(df_X_test)])

# CountVectorizer
f = feature_extraction.text.CountVectorizer()
X_train = f.fit_transform(df_X_train)
X_test = f.transform(df_X_test)

# StandardScaler
scaler = StandardScaler(with_mean=False) # problems with dense matrix
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print([np.shape(X_train), np.shape(X_test)])
print(type(X_train))

[(4173,), (1392,)]
[(4173, 7622), (1392, 7622)]
<class 'scipy.sparse._csr.csr_matrix'>


Let's train Linear SVM on the given Spam/Ham data.


In [ ]:
# SVC

from sklearn.svm import SVC

from sklearn.model_selection import GridSearchCV

params = {
          'C': [1e-4,  0.001, 0.01, 0.1, 1,10] # which hyperparam value of C do you think will work well?
         }

svc = SVC(class_weight={ 0:0.1, 1:0.5 }, kernel='linear')
clf = GridSearchCV(svc, params, scoring = "f1", cv=3)

clf.fit(X_train, y_train)

GridSearchCV(cv=3,
             estimator=SVC(class_weight={0: 0.1, 1: 0.5}, kernel='linear'),
             param_grid={'C': [0.0001, 0.001, 0.01, 0.1, 1, 10]}, scoring='f1')

In [ ]:
res = clf.cv_results_

for i in range(len(res["params"])):
  print(f"Parameters:{res['params'][i]} \n Mean score: {res['mean_test_score'][i]} \n Rank: {res['rank_test_score'][i]}")

Parameters:{'C': 0.0001} 
 Mean score: 0.6566305780023073 
 Rank: 6
Parameters:{'C': 0.001} 
 Mean score: 0.7742322485787693 
 Rank: 1
Parameters:{'C': 0.01} 
 Mean score: 0.767533370474547 
 Rank: 2
Parameters:{'C': 0.1} 
 Mean score: 0.7649416969151316 
 Rank: 3
Parameters:{'C': 1} 
 Mean score: 0.7649416969151316 
 Rank: 3
Parameters:{'C': 10} 
 Mean score: 0.7649416969151316 
 Rank: 3


As you can see,
- we get the best performance when $C=0.001$,
- with F1 Score of 0.77.

\
Now implementing this SVM on the test data.

In [ ]:
svc = SVC(C=0.001,class_weight={ 0:0.1, 1:0.5 }, kernel='linear')

svc.fit(X_train, y_train)

y_pred = svc.predict(X_test)

print(metrics.f1_score(y_test,y_pred))

0.8835820895522388


Linear SVM performs much well
- on the Spam/Ham data
- with F1 Score of 0.88
- when using class weights.